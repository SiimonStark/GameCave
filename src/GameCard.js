import React, { Component } from 'react';

export default class GameCard extends Component {
  constructor(props) {
    super(props);
    this.state= {
      gameTitle: this.props.title,
      gameScore: this.props.metaScore,
      gameGenres: this.props.genres,
      gamePlatforms: this.props.platforms,
      gameMulti: this.props.multiplayer,
      gameImage: this.props.img,
      inLibrary: false
     }
  }
  toggleInfo = () => {
    this.props.setPopup(
      true,
      this.state.gameTitle,
      this.state.gameScore,
      this.state.gameGenres,
      this.state.gamePlatforms,
      this.state.gameMulti,
      this.state.gameImage
    );
  }
  changeLibraryStatus = () => {
    this.setState({ inLibrary: this.state.inLibrary ? false : true });
  }
  render() {
    let libraryButtonClass = "add-to-library " + this.state.inLibrary;
    let inLibraryStatus = " + ";
    inLibraryStatus = this.state.inLibrary ? " - " : " + ";
    const style = { backgroundImage: 'url(' + this.props.img + ')', backgroundSize: 'cover' };
    return(
      <div className="game-card">
        <section className="img" onClick={this.toggleInfo}>
          <div className="gamecover" style={style}></div>
        </section>
        <section className="text-container" onClick={this.toggleInfo}>
          <h4>{this.props.title}</h4>
        </section>
        <section className="btn">
          <button onClick={this.changeLibraryStatus} className={libraryButtonClass}>{inLibraryStatus}</button>
        </section>
      </div>
    )
  }
}