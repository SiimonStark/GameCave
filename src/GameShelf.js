import React, { Component } from 'react';
import GameCard from './GameCard';
import Popup from './Popup';

export default class GameShelf extends Component {
  constructor (props) {
    super (props);
    this.state = {
      popup: false,
      currentTitle: '',
      currentScore: 0,
      currentGenres: [],
      currentPlatforms: [],
      currentMulti: false,
      currentImage: '',
      titleFilter: '',
      genres: [],
      platforms: [],
      isLucky: false,
      isMulti: null
    }
  }
  statePopup = (popup, title, score, genre, plat, multi, img) => {
    this.setState(
      {
        currentGenres: genre,
        currentImage: img,
        currentMulti: multi,
        currentPlatforms: plat,
        currentScore: score,
        currentTitle: title,
        popup: popup
      }
    );
  }
  componentDidUpdate = (prevProps) => {
    if (prevProps !== this.props) {
    this.setState({
      titleFilter: this.props.titleFilter,
      genres: this.props.genresToFilter,
      platforms: this.props.platformsToFilter,
      isLucky: this.props.isLucky,
      isMulti: this.props.multiplayerToggle
    });
  }
}
  filterGamesTitle = (games) => {
    return games.filter(game => game.title.toLowerCase().includes(this.state.titleFilter));
  }

  filterByKey = (games, key) => {
    return games.filter(val => {
        return this.state[key].every(elem => val[key].includes(elem));
    })
  } 
  filterByMulti = (games) => {
    return games.filter(game => game.multiplayer === this.state.isMulti);
  }
  shuffle = (games) => {
    let randoArr = games.sort(() => 0.5 - Math.random());
    return [randoArr[0]];
	}
	
	displayGames = (games) => {
    if (games.length) {
  		games = games.sort((a, b) => a.title.localeCompare(b.title));
  		return games.map((game) => {
  			return (
  				<GameCard {...game}
  					setPopup={this.statePopup}
  					key={game.title} />
  			);
  		});
    } else {
      return (
        <h1 className="empty-shelf">NO RESULTS FOUND</h1>
        )
    }
	}

  render () {
    let games = this.props.games;

    games = this.state.titleFilter ? this.filterGamesTitle(games) : games;
    games = this.state.genres.length ? this.filterByKey(games, 'genres') : games;
    games = this.state.platforms.length ? this.filterByKey(games, 'platforms') : games;
    games = this.state.isMulti !== null ? this.filterByMulti(games) : games;
    games = this.state.isLucky && games.length ? this.shuffle(games) : games;

    const popupOverlay = 
      (this.state.popup && <Popup {...this.state}
        mainGenres={this.props.mainGenres}
        setPopup={this.statePopup}/> )
  return(
    <div className= 'game-shelf'>
      {popupOverlay}
      <article className='card-container'>
        {this.displayGames(games)}
        </article>
      </div>
    )
  }
}