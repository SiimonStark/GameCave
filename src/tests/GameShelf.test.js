import React from 'react';
import ReactDOM from 'react-dom';
import GameShelf from '../GameShelf';
import GameCard from '../GameCard';
import { shallow } from 'enzyme';

const games =[{"title": "The Last of Us",
"platforms": ["Playstation 3", "Playstation 4"],
"genres": ["Action", "Adventure", "Horror"],
"multiplayer": false,
"metaScore": 95,
"img": "https://i.ebayimg.com/images/g/ZX4AAOSw1vlUrlTm/s-l300.jpg"
},
{
  "title": "Super Smash Bros. Ultimate",
  "platforms": ["Switch"],
  "genres": ["Beat-'Em-Up", "Fighting", "Couch-Co-Op"],
  "multiplayer": false,
  "metaScore": 93,
  "img": "https://upload.wikimedia.org/wikipedia/en/thumb/5/50/Super_Smash_Bros._Ultimate.jpg/220px-Super_Smash_Bros._Ultimate.jpg"
},
{
  "title": "The Binding of Isaac: Rebirth",
  "platforms": ["Playstation 4", "Xbox One"],
  "genres": ["Rogue-like", "Horror", "Strategy"],
  "multiplayer": false,
  "metaScore": 83,
  "img": "https://static.giantbomb.com/uploads/scale_small/8/87790/2768049-box_tobir.jpg"
},
{
  "title": "Dark Souls",
  "platforms": ["PC", "Playstation 4", "Xbox One"],
  "genres": ["Action", "Adventure", "RPG", "Fantasy"],
  "multiplayer": true,
  "metaScore": 64,
  "img": "https://upload.wikimedia.org/wikipedia/en/thumb/8/8d/Dark_Souls_Cover_Art.jpg/220px-Dark_Souls_Cover_Art.jpg"}]
  
  const gameMock = {
    "title": "Dark Souls",
    "platforms": ["PC", "Playstation 4", "Xbox One"],
    "genres": ["Action", "Adventure", "RPG", "Fantasy"],
    "multiplayer": true,
    "metaScore": 64,
    "img": "https://upload.wikimedia.org/wikipedia/en/thumb/8/8d/Dark_Souls_Cover_Art.jpg/220px-Dark_Souls_Cover_Art.jpg"
  };
  
  describe('GameShelf', () => {
    let wrapper;
    beforeEach(() => {
      wrapper = shallow(
        <GameShelf games={games}/>
        );
      });
      it('should match snapshot', () => {
        expect(wrapper).toMatchSnapshot();
      });
      it ('should have default state', () => {
        expect(wrapper.state()).toEqual({ 
          popup: false,
          currentTitle: '',
          currentScore: 0,
          currentGenres: [],
          currentPlatforms: [],
          isLucky: false,
          isMulti: null,
          currentMulti: false,
          currentImage: '',
          titleFilter: '',
          genres: [],
          platforms: []
        });
      });
      
      it('should update states after invoking statePopup', () => {
        wrapper.instance().statePopup(
          true,
          games[0].title,
          games[0].metaScore,
          games[0].genres,
          games[0].platforms,
          games[0].multiplayer,
          games[0].img)
          expect(wrapper.state()).toEqual({
            popup: true,
            currentTitle: games[0].title,
            currentScore: games[0].metaScore,
            currentGenres: games[0].genres,
            currentPlatforms: games[0].platforms,
            isLucky: false,
            isMulti: null,
            currentMulti: games[0].multiplayer,
            currentImage: games[0].img,
            titleFilter: '',
            genres: [],
            platforms: []
          });
        });
        
        it('should be able to filter games by title', () => {
          wrapper.setState({ titleFilter: 'dark'});
          const results = wrapper.instance().filterGamesTitle(games);
          expect(results).toEqual([gameMock]);
        });
        
        it('should be able to filter games by key', () => {
          wrapper.setState({platforms: ["PC"]});
          const results = wrapper.instance().filterByKey(games, "platforms");
          expect(results).toEqual([gameMock]);
        });
        
        it('should be able to filter games by multiplayer value', () => {
          wrapper.setState({isMulti: true});
          const results = wrapper.instance().filterByMulti(games);
          expect(results).toEqual([gameMock]);
        });
        
        it('when shuffle is invoked games array should be at a length of 1', () => {
          expect(games.length).toEqual(4)
          wrapper.setState({ isLucky: true });
          const results = wrapper.instance().shuffle(games);
          expect(results.length).toEqual(1); 
        });
      });