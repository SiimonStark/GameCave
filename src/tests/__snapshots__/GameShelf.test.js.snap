// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameShelf should match snapshot 1`] = `
<div
  className="game-shelf"
>
  <article
    className="card-container"
  >
    <GameCard
      genres={
        Array [
          "Action",
          "Adventure",
          "RPG",
          "Fantasy",
        ]
      }
      img="https://upload.wikimedia.org/wikipedia/en/thumb/8/8d/Dark_Souls_Cover_Art.jpg/220px-Dark_Souls_Cover_Art.jpg"
      key="Dark Souls"
      metaScore={64}
      multiplayer={true}
      platforms={
        Array [
          "PC",
          "Playstation 4",
          "Xbox One",
        ]
      }
      setPopup={[Function]}
      title="Dark Souls"
    />
    <GameCard
      genres={
        Array [
          "Beat-'Em-Up",
          "Fighting",
          "Couch-Co-Op",
        ]
      }
      img="https://upload.wikimedia.org/wikipedia/en/thumb/5/50/Super_Smash_Bros._Ultimate.jpg/220px-Super_Smash_Bros._Ultimate.jpg"
      key="Super Smash Bros. Ultimate"
      metaScore={93}
      multiplayer={false}
      platforms={
        Array [
          "Switch",
        ]
      }
      setPopup={[Function]}
      title="Super Smash Bros. Ultimate"
    />
    <GameCard
      genres={
        Array [
          "Rogue-like",
          "Horror",
          "Strategy",
        ]
      }
      img="https://static.giantbomb.com/uploads/scale_small/8/87790/2768049-box_tobir.jpg"
      key="The Binding of Isaac: Rebirth"
      metaScore={83}
      multiplayer={false}
      platforms={
        Array [
          "Playstation 4",
          "Xbox One",
        ]
      }
      setPopup={[Function]}
      title="The Binding of Isaac: Rebirth"
    />
    <GameCard
      genres={
        Array [
          "Action",
          "Adventure",
          "Horror",
        ]
      }
      img="https://i.ebayimg.com/images/g/ZX4AAOSw1vlUrlTm/s-l300.jpg"
      key="The Last of Us"
      metaScore={95}
      multiplayer={false}
      platforms={
        Array [
          "Playstation 3",
          "Playstation 4",
        ]
      }
      setPopup={[Function]}
      title="The Last of Us"
    />
  </article>
</div>
`;
